src/Tools/Common/ReversedServerHelpers/ReversedServerHelpers.cs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Internal.Common.Utils
+{
+    // <summary>
+    // ProcessLauncher is a child-process launcher for "diagnostics tools at startup" scenarios
+    // It launches the target process at startup and passes its processId to the corresponding Command handler.
+    // </summary>
+    internal class ProcessLauncher
+    {
+        private Process _childProc;
+        public ManualResetEvent HasExited;
+
+        internal static ProcessLauncher Launcher = new ProcessLauncher();
+
+        public void PrepareChildProcess(List<string> args)
+        {
+            _childProc = new Process();
+            _childProc.StartInfo.FileName = args[0];
+            _childProc.StartInfo.Arguments = String.Join(" ", args.GetRange(1, args.Count - 1));
+        }
+
+        public bool HasChildProc
+        {
+            get
+            {
+                return _childProc != null;
+            }
+        }
+
+        public Process ChildProc
+        {
+            get
+            {
+                return _childProc;
+            }
+        }
+        public bool Start(string diagnosticTransportName)
+        {
+            HasExited = new ManualResetEvent(false);
+            _childProc.StartInfo.UseShellExecute = false;
+            _childProc.StartInfo.RedirectStandardOutput = true;
+            _childProc.StartInfo.RedirectStandardError = true;
+            _childProc.StartInfo.RedirectStandardInput = true;
+            _childProc.StartInfo.Environment.Add("DOTNET_DiagnosticPorts", $"{diagnosticTransportName},suspend");
+            _childProc.Exited += new EventHandler(OnExited);
+            try
+            {
+                _childProc.Start();
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine($"Cannot start target process: {_childProc.StartInfo.FileName}");
+                Console.WriteLine(e.ToString());
+                return false;
+            }
+            return true;
+        }
+
+        private static void OnExited(object sender, EventArgs args)
+        {
+            ProcessLauncher.Launcher.HasExited.Set();
+        }
+    }
+
+    // <summary>
+    // This class acts a helper class for building a DiagnosticsClient instance
+    // </summary>
+    internal class ReversedDiagnosticsClientBuilder
+    {
+        private static string GetRandomTransportName() => "DOTNET_TOOL_PATH" + Path.GetRandomFileName();
```

Comment:
Won't this create a path like "DOTNET_TOOL_PATHsagnakg.cdb"? It seems like you meant to do environment variable expansion and Path.Join here

Reply:
It's actually file name. I did want to do a random file name that is prefixed with a well-known string. Maybe I could put a _ between the DOTNET_TOOL_PATH and the random string.

Reply:
It seems confusing to use something with PATH in the name that looks like an environment variable if you just want an identifier. If it were me I would either just use DOTNET_TOOL or do something like DOTNET_TOOL_REVERSE_TRANSPORT to be more descriptive